File: src/McaDiff/Repo/Repository.cs
namespace McaDiff.Repo;

/// <summary>
/// An external, bare repository: the object store plus refs and HEAD. Worlds are
/// committed into it and checked out from it. HEAD names the current branch, which
/// commit/merge advance.
/// </summary>
public sealed class Repository
{
    public const string DefaultBranch = "main";

    public string Dir { get; }
    public ObjectStore Objects { get; }

    private ChunkCache? _cache;
    /// <summary>Lazily-loaded decode cache (see <see cref="ChunkCache"/>).</summary>
    public ChunkCache Cache => _cache ??= ChunkCache.Load(Dir);

    private Repository(string dir)
    {
        Dir = dir;
        Objects = new ObjectStore(dir);
    }

    public static bool IsRepository(string dir) => File.Exists(Path.Combine(dir, "HEAD"));

    public static Repository Init(string dir)
    {
        if (IsRepository(dir))
            throw new InvalidOperationException($"already a repository: {dir}");
        Directory.CreateDirectory(Path.Combine(dir, "objects"));
        Directory.CreateDirectory(Path.Combine(dir, "refs", "heads"));
        File.WriteAllText(Path.Combine(dir, "HEAD"), $"ref: refs/heads/{DefaultBranch}\n");
        return new Repository(dir);
    }

    public static Repository Open(string dir)
    {
        if (!IsRepository(dir))
            throw new InvalidOperationException($"not a repository: {dir}");
        return new Repository(dir);
    }

    /// <summary>
    /// Locates the repository git-style: an explicit <paramref name="dashC"/> if
    /// given, otherwise the current directory or the nearest ancestor that is a
    /// repository. Returns null if none is found.
    /// </summary>
    public static Repository? Discover(string? dashC)
    {
        if (dashC is not null)
            return IsRepository(dashC) ? Open(dashC) : null;
        for (string? dir = Directory.GetCurrentDirectory(); dir is not null; dir = Directory.GetParent(dir)?.FullName)
            if (IsRepository(dir))
                return Open(dir);
        return null;
    }

    // ---- config (bound worktree) ----

    /// <summary>The world directory bound to this repo (git's "working tree"), or null.</summary>
    public string? Worktree
    {
        get => ReadConfig().Worktree;
        set { RepoConfig c = ReadConfig(); c.Worktree = value is null ? null : Path.GetFullPath(value); WriteConfig(c); }
    }

    private string ConfigPath => Path.Combine(Dir, "config");
    private RepoConfig ReadConfig() => File.Exists(ConfigPath)
        ? System.Text.Json.JsonSerializer.Deserialize<RepoConfig>(File.ReadAllText(ConfigPath)) ?? new RepoConfig()
        : new RepoConfig();
    private void WriteConfig(RepoConfig c) => File.WriteAllText(ConfigPath,
        System.Text.Json.JsonSerializer.Serialize(c, new System.Text.Json.JsonSerializerOptions { WriteIndented = true }));

    private sealed class RepoConfig { public string? Worktree { get; set; } }

    // ---- HEAD & branches ----

    public string ReadHeadRaw() => File.ReadAllText(Path.Combine(Dir, "HEAD")).Trim();

    /// <summary>Current branch name, or null if HEAD is detached at a commit.</summary>
    public string? CurrentBranch()
    {
        string head = ReadHeadRaw();
        return head.StartsWith("ref: refs/heads/", StringComparison.Ordinal)
            ? head["ref: refs/heads/".Length..]
            : null;
    }

    public void SetHeadToBranch(string branch) =>
        File.WriteAllText(Path.Combine(Dir, "HEAD"), $"ref: refs/heads/{branch}\n");

    public void SetHeadDetached(string commitHash) =>
        File.WriteAllText(Path.Combine(Dir, "HEAD"), commitHash + "\n");

    public string? ReadBranch(string branch)
    {
        string p = BranchPath(branch);
        return File.Exists(p) ? File.ReadAllText(p).Trim() : null;
    }

    public void WriteBranch(string branch, string commitHash)
    {
        string p = BranchPath(branch);
        Directory.CreateDirectory(Path.GetDirectoryName(p)!);
        File.WriteAllText(p, commitHash + "\n");
    }

    public IEnumerable<string> Branches()
    {
        string dir = Path.Combine(Dir, "refs", "heads");
        if (!Directory.Exists(dir)) yield break;
        foreach (string f in Directory.EnumerateFiles(dir).OrderBy(x => x, StringComparer.Ordinal))
            yield return Path.GetFileName(f);
    }

    /// <summary>The commit HEAD currently points at (via its branch), or null if unborn.</summary>
    public string? HeadCommit()
    {
        string? branch = CurrentBranch();
        return branch is not null ? ReadBranch(branch) : ReadHeadRaw();
    }

    /// <summary>Resolves <c>HEAD</c>, a branch name, or a commit hash to a commit hash.</summary>
    public string ResolveRef(string refName)
    {
        if (refName == "HEAD")
            return HeadCommit() ?? throw new InvalidOperationException("HEAD has no commits yet");
        if (ReadBranch(refName) is { } byBranch) return byBranch;
        if (Objects.Exists(refName)) return refName;
        throw new InvalidOperationException($"unknown ref: {refName}");
    }

    // ---- typed object IO ----

    public CommitObject ReadCommit(string hash) => CommitObject.FromJson(Objects.ReadText(hash));
    public string WriteCommit(CommitObject c) => Objects.WriteText(c.ToJson());
    public Manifest ReadManifest(string hash) => Manifest.FromJson(Objects.ReadText(hash));
    public string WriteManifest(Manifest m) => Objects.WriteText(m.ToJson());

    /// <summary>
    /// Writes a commit object and advances the current branch (creating it if this
    /// is the first commit). Returns the new commit hash.
    /// </summary>
    public string CreateCommit(string treeHash, IReadOnlyList<string> parents, string message, string author)
    {
        var commit = new CommitObject
        {
            Tree = treeHash,
            Parents = [.. parents],
            Message = message,
            Author = author,
            Time = DateTimeOffset.Now.ToString("o"),
        };
        string hash = WriteCommit(commit);

        // Advance the current branch; if HEAD is detached, move HEAD itself and
        // leave every branch untouched (committing on a detached HEAD must never
        // silently clobber 'main').
        if (CurrentBranch() is { } branch) WriteBranch(branch, hash);
        else SetHeadDetached(hash);
        return hash;
    }

    private string BranchPath(string branch) => Path.Combine(Dir, "refs", "heads", branch);
}

File: rust/crates/repo/src/repository.rs
//! The bare, external repository: object store + refs/HEAD + config, plus
//! commit/manifest storage and revision resolution.

use crate::manifest::{CommitObject, Manifest};
use crate::object_store::ObjectStore;
use crate::{RepoError, Result};
use std::path::{Path, PathBuf};

pub struct Repository {
    dir: PathBuf,
    objects: ObjectStore,
}

impl Repository {
    pub fn dir(&self) -> &Path {
        &self.dir
    }
    pub fn objects(&self) -> &ObjectStore {
        &self.objects
    }

    /// A directory is a repo if it has both `HEAD` and `objects/`.
    pub fn is_repository(dir: &Path) -> bool {
        dir.join("HEAD").is_file() && dir.join("objects").is_dir()
    }

    /// Create a fresh repo (idempotent — re-init keeps existing HEAD).
    pub fn init(dir: &Path) -> Result<Self> {
        std::fs::create_dir_all(dir.join("objects"))?;
        std::fs::create_dir_all(dir.join("refs").join("heads"))?;
        let head = dir.join("HEAD");
        if !head.exists() {
            std::fs::write(&head, "ref: refs/heads/main\n")?;
        }
        Self::open(dir)
    }

    pub fn open(dir: &Path) -> Result<Self> {
        if !Self::is_repository(dir) {
            return Err(RepoError::NotARepository(dir.display().to_string()));
        }
        Ok(Self {
            dir: dir.to_path_buf(),
            objects: ObjectStore::new(dir.join("objects")),
        })
    }

    /// Walk up from `start` looking for a repo directory (git-style).
    pub fn discover(start: &Path) -> Result<Self> {
        let mut cur = std::fs::canonicalize(start).unwrap_or_else(|_| start.to_path_buf());
        loop {
            if Self::is_repository(&cur) {
                return Self::open(&cur);
            }
            match cur.parent() {
                Some(p) => cur = p.to_path_buf(),
                None => return Err(RepoError::NotARepository(start.display().to_string())),
            }
        }
    }

    // ---- config (simple `key = value` lines) ----

    fn config_path(&self) -> PathBuf {
        self.dir.join("config")
    }

    pub fn config_get(&self, key: &str) -> Option<String> {
        let text = std::fs::read_to_string(self.config_path()).ok()?;
        for line in text.lines() {
            if let Some((k, v)) = line.split_once('=') {
                if k.trim() == key {
                    return Some(v.trim().to_string());
                }
            }
        }
        None
    }

    pub fn config_set(&self, key: &str, value: &str) -> Result<()> {
        let mut lines: Vec<String> = std::fs::read_to_string(self.config_path())
            .unwrap_or_default()
            .lines()
            .filter(|l| l.split_once('=').map(|(k, _)| k.trim() != key).unwrap_or(true))
            .map(|l| l.to_string())
            .collect();
        lines.push(format!("{key} = {value}"));
        std::fs::write(self.config_path(), lines.join("\n") + "\n")?;
        Ok(())
    }

    pub fn worktree(&self) -> Option<String> {
        self.config_get("worktree")
    }
    pub fn set_worktree(&self, w: &str) -> Result<()> {
        self.config_set("worktree", w)
    }

    // ---- refs / HEAD ----

    fn head_path(&self) -> PathBuf {
        self.dir.join("HEAD")
    }
    fn branch_path(&self, name: &str) -> PathBuf {
        self.dir.join("refs").join("heads").join(name)
    }

    pub fn read_branch(&self, name: &str) -> Option<String> {
        std::fs::read_to_string(self.branch_path(name))
            .ok()
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
    }

    pub fn write_branch(&self, name: &str, hash: &str) -> Result<()> {
        let p = self.branch_path(name);
        if let Some(parent) = p.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::write(p, format!("{hash}\n"))?;
        Ok(())
    }

    pub fn branches(&self) -> Vec<String> {
        let mut out = Vec::new();
        if let Ok(entries) = std::fs::read_dir(self.dir.join("refs").join("heads")) {
            for e in entries.flatten() {
                if let Ok(name) = e.file_name().into_string() {
                    out.push(name);
                }
            }
        }
        out.sort();
        out
    }

    pub fn current_branch(&self) -> Option<String> {
        let head = std::fs::read_to_string(self.head_path()).ok()?;
        head.trim()
            .strip_prefix("ref: refs/heads/")
            .map(|s| s.to_string())
    }

    pub fn set_head_to_branch(&self, name: &str) -> Result<()> {
        std::fs::write(self.head_path(), format!("ref: refs/heads/{name}\n"))?;
        Ok(())
    }

    pub fn set_head_detached(&self, hash: &str) -> Result<()> {
        std::fs::write(self.head_path(), format!("{hash}\n"))?;
        Ok(())
    }

    pub fn head_commit(&self) -> Option<String> {
        let head = std::fs::read_to_string(self.head_path()).ok()?;
        let head = head.trim();
        if let Some(branch) = head.strip_prefix("ref: refs/heads/") {
            self.read_branch(branch)
        } else if head.is_empty() {
            None
        } else {
            Some(head.to_string())
        }
    }

    // ---- objects: commits + manifests ----

    pub fn write_manifest(&self, m: &Manifest) -> Result<String> {
        self.objects.write(m.to_json()?.as_bytes())
    }
    pub fn read_manifest(&self, tree: &str) -> Result<Manifest> {
        Manifest::from_json(&String::from_utf8_lossy(&self.objects.read(tree)?))
    }
    pub fn write_commit(&self, c: &CommitObject) -> Result<String> {
        self.objects.write(c.to_json()?.as_bytes())
    }
    pub fn read_commit(&self, hash: &str) -> Result<CommitObject> {
        CommitObject::from_json(&String::from_utf8_lossy(&self.objects.read(hash)?))
    }

    pub fn create_commit(
        &self,
        tree: &str,
        parents: Vec<String>,
        message: &str,
        author: &str,
        time: &str,
    ) -> Result<String> {
        self.write_commit(&CommitObject {
            tree: tree.to_string(),
            parents,
            message: message.to_string(),
            author: author.to_string(),
            time: time.to_string(),
            committer: None,
            commit_time: None,
        })
    }

    // ---- revision resolution ----

    /// Resolve a revision spec: `HEAD`, a branch name, a (possibly abbreviated)
    /// hex id, with optional trailing `~n` (n-th first-parent ancestor) and
    /// `^`/`^n` (n-th parent).
    pub fn resolve_ref(&self, spec: &str) -> Result<String> {
        let split = spec.find(['~', '^']).unwrap_or(spec.len());
        let base = &spec[..split];
        let mut hash = self.resolve_base(base)?;
        let mut rest = &spec[split..];
        while let Some(op) = rest.chars().next() {
            rest = &rest[op.len_utf8()..];
            let digits = rest.bytes().take_while(u8::is_ascii_digit).count();
            let n: usize = if digits > 0 {
                rest[..digits].parse().unwrap_or(1)
            } else {
                1
            };
            rest = &rest[digits..];
            hash = match op {
                '~' => {
                    let mut h = hash;
                    for _ in 0..n {
                        h = self.first_parent(&h)?;
                    }
                    h
                }
                '^' => self.nth_parent(&hash, n)?,
                _ => return Err(RepoError::BadRef(spec.to_string())),
            };
        }
        Ok(hash)
    }

    fn resolve_base(&self, base: &str) -> Result<String> {
        if base.is_empty() || base == "HEAD" {
            return self
                .head_commit()
                .ok_or_else(|| RepoError::BadRef("HEAD".into()));
        }
        if let Some(h) = self.read_branch(base) {
            return Ok(h);
        }
        if base.len() >= 4 && base.chars().all(|c| c.is_ascii_hexdigit()) {
            if self.objects.exists(base) {
                return Ok(base.to_string());
            }
            if let Some(full) = self.find_object_by_prefix(base) {
                return Ok(full);
            }
        }
        Err(RepoError::BadRef(base.to_string()))
    }

    fn first_parent(&self, h: &str) -> Result<String> {
        self.read_commit(h)?
            .parents
            .into_iter()
            .next()
            .ok_or_else(|| RepoError::BadRef(format!("{h}: no parent")))
    }

    fn nth_parent(&self, h: &str, n: usize) -> Result<String> {
        self.read_commit(h)?
            .parents
            .get(n.saturating_sub(1))
            .cloned()
            .ok_or_else(|| RepoError::BadRef(format!("{h}^{n}: no such parent")))
    }

    fn find_object_by_prefix(&self, prefix: &str) -> Option<String> {
        let (sub, rest) = prefix.split_at(2);
        let entries = std::fs::read_dir(self.dir.join("objects").join(sub)).ok()?;
        let mut found = None;
        for e in entries.flatten() {
            let name = e.file_name().to_string_lossy().to_string();
            if name.ends_with(".tmp") || !name.starts_with(rest) {
                continue;
            }
            if found.is_some() {
                return None; // ambiguous
            }
            found = Some(format!("{sub}{name}"));
        }
        found
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn init_open_and_config() {
        let d = tempfile::tempdir().unwrap();
        let repo = Repository::init(d.path()).unwrap();
        assert!(Repository::is_repository(d.path()));
        repo.set_worktree("/some/world").unwrap();
        assert_eq!(repo.worktree().as_deref(), Some("/some/world"));
        // re-open sees the config
        let repo2 = Repository::open(d.path()).unwrap();
        assert_eq!(repo2.worktree().as_deref(), Some("/some/world"));
    }

    #[test]
    fn commit_flow_and_resolution() {
        let d = tempfile::tempdir().unwrap();
        let repo = Repository::init(d.path()).unwrap();
        let tree = repo.write_manifest(&Manifest::default()).unwrap();

        let c1 = repo.create_commit(&tree, vec![], "first", "me", "t1").unwrap();
        repo.write_branch("main", &c1).unwrap();
        assert_eq!(repo.head_commit().as_deref(), Some(c1.as_str()));
        assert_eq!(repo.current_branch().as_deref(), Some("main"));

        let c2 = repo
            .create_commit(&tree, vec![c1.clone()], "second", "me", "t2")
            .unwrap();
        repo.write_branch("main", &c2).unwrap();

        assert_eq!(repo.resolve_ref("HEAD").unwrap(), c2);
        assert_eq!(repo.resolve_ref("main").unwrap(), c2);
        assert_eq!(repo.resolve_ref("HEAD~1").unwrap(), c1);
        assert_eq!(repo.resolve_ref("HEAD^").unwrap(), c1);
        assert_eq!(repo.resolve_ref(&c1).unwrap(), c1);
        assert_eq!(repo.resolve_ref(&c1[..8]).unwrap(), c1); // abbreviated
        assert!(repo.resolve_ref("nope").is_err());

        let commit = repo.read_commit(&c2).unwrap();
        assert_eq!(commit.message, "second");
        assert_eq!(commit.parents, vec![c1]);
    }

    #[test]
    fn discover_walks_up() {
        let d = tempfile::tempdir().unwrap();
        Repository::init(d.path()).unwrap();
        let nested = d.path().join("a").join("b");
        std::fs::create_dir_all(&nested).unwrap();
        let repo = Repository::discover(&nested).unwrap();
        assert!(Repository::is_repository(repo.dir()));
    }
}
